solidity/test/bridge/WalletCoordinator.test.ts

```diff
@@ -0,0 +1,921 @@
+import { ethers, helpers } from "hardhat"
+import chai, { expect } from "chai"
+import { FakeContract, smock } from "@defi-wonderland/smock"
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import { BigNumber, ContractTransaction } from "ethers"
+import type {
+  WalletCoordinator,
+  Bridge,
+  ReimbursementPool,
+  IWalletRegistry,
+} from "../../typechain"
+
+chai.use(smock.matchers)
+
+const { lastBlockTime, increaseTime } = helpers.time
+const { createSnapshot, restoreSnapshot } = helpers.snapshot
+const { AddressZero, HashZero } = ethers.constants
+
+const emptyWalletMemberContext = {
+  walletMembersIDs: [],
+  walletMemberIndex: 0,
+}
+
+describe("WalletCoordinator", () => {
+  let owner: SignerWithAddress
+  let thirdParty: SignerWithAddress
+
+  let walletRegistry: FakeContract<IWalletRegistry>
+  let bridge: FakeContract<Bridge>
+  let reimbursementPool: FakeContract<ReimbursementPool>
+
+  let walletCoordinator: WalletCoordinator
+
+  before(async () => {
+    // eslint-disable-next-line @typescript-eslint/no-extra-semi
+    ;[owner, thirdParty] = await helpers.signers.getUnnamedSigners()
+
+    walletRegistry = await smock.fake<IWalletRegistry>("IWalletRegistry")
+
+    bridge = await smock.fake<Bridge>("Bridge")
+
+    bridge.contractReferences.returns([
+      AddressZero,
+      AddressZero,
+      walletRegistry.address,
+      AddressZero,
+    ])
+
+    reimbursementPool = await smock.fake<ReimbursementPool>("ReimbursementPool")
+
+    const WalletCoordinator = await ethers.getContractFactory(
+      "WalletCoordinator"
+    )
+
+    walletCoordinator = await WalletCoordinator.deploy()
+
+    await walletCoordinator.initialize(
+      bridge.address,
+      reimbursementPool.address
+    )
+  })
+
+  describe("addProposalSubmitter", () => {
+    before(async () => {
+      await createSnapshot()
+    })
+
+    after(async () => {
+      await restoreSnapshot()
+    })
```

Comment:
Are they needed if we create and restore snapshot for every test modifying the state? Not a big deal but with the number of tests we have, doing and restoring snapshots only when necessary could potentially help with `yarn test` execution time.

Reply:
Honestly, I'd keep it as is because:
- The execution time is negligible (around 1 millisecond)
- That helps avoid future debugging in case new state modifications are added at the top level of the scenario. It happened to me to forget the snapshot instructions and lose time in finding the cause of the weird behavior of specific scenarios.

Reply:
Works for me!